File: src/HVM/Reduce.hs
{-./Type.hs-}
{-./Foreign.hs-}

-- NOTE: THIS FILE IS MISSING HTE INC/DEC INTERACTIONS. LET'S FIX IT

module HVM.Reduce where

import Control.Monad (when, forM, forM_)
import Data.Word
import HVM.Collapse
import HVM.Extract
import HVM.Foreign
import HVM.Inject
import HVM.Type
import System.Exit
import qualified Data.Map.Strict as MS

reduceAt :: Bool -> ReduceAt

reduceAt debug book host = do 
  term <- got host
  let tag = termTag term
  let lab = termLab term
  let loc = termLoc term

  when debug $ do
    root <- doExtractCoreAt gotT book 0
    core <- doExtractCoreAt gotT book host
    putStrLn $ "reduce: " ++ showTerm term
    -- putStrLn $ "---------------- CORE: "
    -- putStrLn $ showCore core
    putStrLn $ "---------------- ROOT: "
    putStrLn $ showCore (doLiftDups root)

  case tag of
    t | t == _LET_ -> do
      case modeT lab of
        LAZY -> do
          val <- got (loc + 0)
          cont host (reduceLet term val)
        STRI -> do
          val <- reduceAt debug book (loc + 0)
          cont host (reduceLet term val)

    t | t == _APP_ -> do
      fun <- reduceAt debug book (loc + 0)
      case termTag fun of
        t | t == _ERA_ -> cont host (reduceAppEra term fun)
        t | t == _LAM_ -> cont host (reduceAppLam term fun)
        t | t == _SUP_ -> cont host (reduceAppSup term fun)
        t | t == _CTR_ -> cont host (reduceAppCtr term fun)
        t | t == _W32_ -> cont host (reduceAppW32 term fun)
        t | t == _CHR_ -> cont host (reduceAppW32 term fun)
        t | t == _INC_ -> cont host (reduceAppInc term fun)
        t | t == _DEC_ -> cont host (reduceAppDec term fun)
        _   -> set (loc + 0) fun >> return term

    t | t == _MAT_ -> do
      val <- reduceAt debug book (loc + 0)
      case termTag val of
        t | t == _ERA_ -> cont host (reduceMatEra term val)
        t | t == _LAM_ -> cont host (reduceMatLam term val)
        t | t == _SUP_ -> cont host (reduceMatSup term val)
        t | t == _CTR_ -> cont host (reduceMatCtr term val)
        t | t == _W32_ -> cont host (reduceMatW32 term val)
        t | t == _CHR_ -> cont host (reduceMatW32 term val)
        t | t == _INC_ -> cont host (reduceMatInc term val)
        t | t == _DEC_ -> cont host (reduceMatDec term val)
        _   -> set (loc + 0) val >> return term

    t | t == _IFL_ -> do
      val <- reduceAt debug book (loc + 0)
      case termTag val of
        t | t == _ERA_ -> cont host (reduceMatEra term val)
        t | t == _LAM_ -> cont host (reduceMatLam term val)
        t | t == _SUP_ -> cont host (reduceMatSup term val)
        t | t == _CTR_ -> cont host (reduceMatCtr term val)
        t | t == _W32_ -> cont host (reduceMatW32 term val)
        t | t == _CHR_ -> cont host (reduceMatW32 term val)
        t | t == _INC_ -> cont host (reduceMatInc term val)
        t | t == _DEC_ -> cont host (reduceMatDec term val)
        _   -> set (loc + 0) val >> return term

    t | t == _SWI_ -> do
      val <- reduceAt debug book (loc + 0)
      case termTag val of
        t | t == _ERA_ -> cont host (reduceMatEra term val)
        t | t == _LAM_ -> cont host (reduceMatLam term val)
        t | t == _SUP_ -> cont host (reduceMatSup term val)
        t | t == _CTR_ -> cont host (reduceMatCtr term val)
        t | t == _W32_ -> cont host (reduceMatW32 term val)
        t | t == _CHR_ -> cont host (reduceMatW32 term val)
        t | t == _INC_ -> cont host (reduceMatInc term val)
        t | t == _DEC_ -> cont host (reduceMatDec term val)
        _   -> set (loc + 0) val >> return term

    t | t == _OPX_ -> do
      val <- reduceAt debug book (loc + 0)
      case termTag val of
        t | t == _ERA_ -> cont host (reduceOpxEra term val)
        t | t == _LAM_ -> cont host (reduceOpxLam term val)
        t | t == _SUP_ -> cont host (reduceOpxSup term val)
        t | t == _CTR_ -> cont host (reduceOpxCtr term val)
        t | t == _W32_ -> cont host (reduceOpxW32 term val)
        t | t == _CHR_ -> cont host (reduceOpxW32 term val)
        t | t == _INC_ -> cont host (reduceOpxInc term val)
        t | t == _DEC_ -> cont host (reduceOpxDec term val)
        _   -> set (loc + 0) val >> return term

    t | t == _OPY_ -> do
      val <- reduceAt debug book (loc + 0)
      case termTag val of
        t | t == _ERA_ -> cont host (reduceOpyEra term val)
        t | t == _LAM_ -> cont host (reduceOpyLam term val)
        t | t == _SUP_ -> cont host (reduceOpySup term val)
        t | t == _CTR_ -> cont host (reduceOpyCtr term val)
        t | t == _W32_ -> cont host (reduceOpyW32 term val)
        t | t == _CHR_ -> cont host (reduceOpyW32 term val)
        t | t == _INC_ -> cont host (reduceOpyInc term val)
        t | t == _DEC_ -> cont host (reduceOpyDec term val)
        _   -> set (loc + 0) val >> return term

    t | t == _DP0_ -> do
      sb0 <- got (loc + 0)
      if termGetBit sb0 == 0
        then do
          val <- reduceAt debug book (loc + 0)
          case termTag val of
            t | t == _ERA_ -> cont host (reduceDupEra term val)
            t | t == _LAM_ -> cont host (reduceDupLam term val)
            t | t == _SUP_ -> cont host (reduceDupSup term val)
            t | t == _CTR_ -> cont host (reduceDupCtr term val)
            t | t == _W32_ -> cont host (reduceDupW32 term val)
            t | t == _CHR_ -> cont host (reduceDupW32 term val)
            t | t == _INC_ -> cont host (reduceDupInc term val)
            t | t == _DEC_ -> cont host (reduceDupDec term val)
            _   -> set (loc + 0) val >> return term
        else do
          set host (termRemBit sb0)
          freeNode (loc + 0) 1 -- sub cell consumed by its last occurrence
          reduceAt debug book host

    t | t == _DP1_ -> do
      sb1 <- got (loc + 0)
      if termGetBit sb1 == 0
        then do
          val <- reduceAt debug book (loc + 0)
          case termTag val of
            t | t == _ERA_ -> cont host (reduceDupEra term val)
            t | t == _LAM_ -> cont host (reduceDupLam term val)
            t | t == _SUP_ -> cont host (reduceDupSup term val)
            t | t == _CTR_ -> cont host (reduceDupCtr term val)
            t | t == _W32_ -> cont host (reduceDupW32 term val)
            t | t == _CHR_ -> cont host (reduceDupW32 term val)
            t | t == _INC_ -> cont host (reduceDupInc term val)
            t | t == _DEC_ -> cont host (reduceDupDec term val)
            _   -> set (loc + 0) val >> return term
        else do
          set host (termRemBit sb1)
          freeNode (loc + 0) 1 -- sub cell consumed by its last occurrence
          reduceAt debug book host

    t | t == _VAR_ -> do
      sub <- got (loc + 0)
      if termGetBit sub == 0
        then return term
        else do
          set host (termRemBit sub)
          freeNode (loc + 0) 1 -- sub cell consumed by its VAR occurrence
          reduceAt debug book host

    t | t == _REF_ -> do
      reduceRefAt book host
      reduceAt debug book host

    _ -> do
      return term

  where
    cont host action = do
      ret <- action
      set host ret
      reduceAt debug book host

gotT :: Book -> Loc -> HVM Term
gotT book host = got host

reduceRefAt :: Book -> Loc -> HVM Term
reduceRefAt book host = do
  term <- got host
  let lab = termLab term
  let loc = termLoc term
  let fid = fromIntegral lab
  let ari = funArity book fid
  case lab of
    x | x == _DUP_F_ -> reduceRefAt_DupF book host loc ari
    x | x == _SUP_F_ -> reduceRefAt_SupF book host loc ari
    x | x == _LOG_F_ -> reduceRefAt_LogF book host loc ari
    otherwise -> case MS.lookup fid (fidToFun book) of
      Just ((copy, args), core) -> do
        incItr
        when (length args /= fromIntegral ari) $ do
          putStrLn $ "RUNTIME_ERROR: arity mismatch on call to '@" ++ mget (fidToNam book) fid ++ "'."
          exitFailure
        argTerms <- if ari == 0
          then return []
          else forM (zip [0..] args) $ \(i, (strict, _)) -> do
            term <- got (loc + i)
            if strict
              then reduceAt False book (loc + i)
              else return term
        ret <- doInjectCoreAt book core host $ zip (map snd args) argTerms
        -- The REF arg block is dead once the body is injected (interpreted
        -- path only; compiled fast-mode statically reuses this block itself).
        when (ari > 0) $ freeNode loc (fromIntegral ari)
        return ret
      Nothing -> do
        putStrLn $ "RUNTIME_ERROR: Function ID " ++ show fid ++ " not found in fidToFun book."
        exitFailure

-- Primitive: Dynamic Dup `@DUP(lab val λdp0λdp1(bod))`
reduceRefAt_DupF :: Book -> Loc -> Loc -> Word16 -> HVM Term  
reduceRefAt_DupF book host loc ari = do
  incItr
  when (ari /= 3) $ do
    putStrLn $ "RUNTIME_ERROR: arity mismatch on call to '@DUP'."
    exitFailure
  lab <- reduceAt False book (loc + 0)
  val <- got (loc + 1)
  bod <- got (loc + 2)
  dup <- allocNode 1
  case termTag lab of
    t | t == _W32_ -> do
      when (termLoc lab > 0xFFFF) $ do
        error "RUNTIME_ERROR: dynamic DUP label too large"
      -- Create the DUP node with value
      set (dup + 0) val
      -- Create first APP node for (APP bod DP0)
      app1 <- allocNode 2
      set (app1 + 0) bod
      set (app1 + 1) (termNew _DP0_ (fromIntegral (termLoc lab)) dup)
      -- Create second APP node for (APP (APP bod DP0) DP1)
      app2 <- allocNode 2
      set (app2 + 0) (termNew _APP_ 0 app1)
      set (app2 + 1) (termNew _DP1_ (fromIntegral (termLoc lab)) dup)
      let ret = termNew _APP_ 0 app2
      set host ret
      return ret
    _ -> do
      core <- doExtractCoreAt gotT book (loc + 0)
      putStrLn $ "RUNTIME_ERROR: dynamic DUP without numeric label: " ++ showTerm lab
      putStrLn $ showCore (doLiftDups core)
      exitFailure

-- Primitive: Dynamic Sup `@SUP(lab tm0 tm1)`
reduceRefAt_SupF :: Book -> Loc -> Loc -> Word16 -> HVM Term
reduceRefAt_SupF book host loc ari = do
  incItr
  when (ari /= 3) $ do
    putStrLn $ "RUNTIME_ERROR: arity mismatch on call to '@SUP'."
    exitFailure
  lab <- reduceAt False book (loc + 0)
  tm0 <- got (loc + 1)
  tm1 <- got (loc + 2)
  sup <- allocNode 2
  case termTag lab of
    t | t == _W32_ -> do
      when (termLoc lab > 0xFFFF) $ do
        error "RUNTIME_ERROR: dynamic SUP label too large"
      let ret = termNew _SUP_ (fromIntegral (termLoc lab)) sup
      set (sup + 0) tm0
      set (sup + 1) tm1
      set host ret
      return ret
    _ -> error "RUNTIME_ERROR: dynamic SUP without numeric label."

-- Primitive: Logger `@LOG(msg)`
-- Will extract the term and log it. 
-- Returns 0.
reduceRefAt_LogF :: Book -> Loc -> Loc -> Word16 -> HVM Term
reduceRefAt_LogF book host loc ari = do
  incItr
  when (ari /= 1) $ do
    putStrLn $ "RUNTIME_ERROR: arity mismatch on call to '@LOG'."
    exitFailure
  msg <- doExtractCoreAt (reduceAt False) book (loc + 0)
  putStrLn $ showCore msg
  -- msgs <- doCollapseFlatAt gotT book (loc + 0)
  -- forM_ msgs $ \msg -> do
    -- putStrLn $ showCore msg
  let ret = termNew _W32_ 0 0
  set host ret
  return ret

-- Primitive: Fresh `@FRESH`
-- Returns a fresh dup label.
reduceRefAt_FreshF :: Book -> Loc -> Loc -> Word16 -> HVM Term
reduceRefAt_FreshF book host loc ari = do
  incItr
  when (ari /= 0) $ do
    putStrLn $ "RUNTIME_ERROR: arity mismatch on call to '@Fresh'."
    exitFailure
  num <- fromIntegral <$> fresh
  let ret = termNew _W32_ 0 num
  set host ret
  return ret

reduceCAt :: Bool -> ReduceAt
reduceCAt = \ _ _ host -> reduceAtC host

File: src/HVM/runtime/reduce.c
#include "Runtime.h"

// Core reducer and helpers (dispatcher, WHNF, normal form)

static Term reduce_impl(Term term, bool atomic_subs) {
  if (term_tag(term) >= ERA) return term;
  Term  next = term;
  u64   stop = *HVM.spos;
  u64   spos = stop;
  Term* sbuf = HVM.sbuf;

  while (1) {
    Tag tag = term_tag(next);
    Lab lab = term_lab(next);
    Loc loc = term_loc(next);

    // On variables: substitute
    // On eliminators: move to field
    switch (tag) {
      case LET: {
        switch (lab) {
          case LAZY: next = reduce_let(next, got(loc + 0)); continue;
          case STRI: spush(next, sbuf, &spos); next = got(loc + 0); continue;
          default:  printf("invalid:let"); exit(0);
        }
      }
      case APP:
      case MAT:
      case IFL:
      case SWI:
      case OPX:
      case OPY: { spush(next, sbuf, &spos); next = got(loc + 0); continue; }
      case DP0:
      case DP1: {
        Term sub = got(loc + 0);
        if (term_get_bit(sub) == 0) { spush(next, sbuf, &spos); next = sub; continue; }
        if (atomic_subs) sub = take(loc + 0);
        next = term_rem_bit(sub); free_node(loc, 1); continue; // sub cell consumed
      }
      case VAR: {
        Term sub = got(loc);
        if (term_get_bit(sub) == 0) break;
        if (atomic_subs) sub = take(loc);
        next = term_rem_bit(sub); free_node(loc, 1); continue; // sub cell consumed
      }
      case REF: { *HVM.spos = spos; next = reduce_ref(next); spos = *HVM.spos; continue; }
      default: break;
    }

    // Empty stack: term is in WHNF
    if (spos == stop) { *HVM.spos = spos; return next; }

    // Interaction Dispatcher
    Term prev = spop(sbuf, &spos);
    switch (term_tag(prev)) {
      case LET: next = reduce_let(prev, next); continue;
      case APP: switch (tag) {
        case ERA: next = reduce_app_era(prev, next); continue;
        case LAM: next = reduce_app_lam(prev, next); continue;
        case SUP: next = reduce_app_sup(prev, next); continue;
        case CTR: next = reduce_app_ctr(prev, next); continue;
        case W32: case CHR: next = reduce_app_w32(prev, next); continue;
        case INC: next = reduce_app_inc(prev, next); continue;
        case DEC: next = reduce_app_dec(prev, next); continue;
        default: break;
      }
      case DP0:
      case DP1: switch (tag) {
        case ERA: next = reduce_dup_era(prev, next); continue;
        case LAM: next = reduce_dup_lam(prev, next); continue;
        case SUP: next = reduce_dup_sup(prev, next); continue;
        case CTR: next = reduce_dup_ctr(prev, next); continue;
        case W32: case CHR: next = reduce_dup_w32(prev, next); continue;
        case INC: next = reduce_dup_inc(prev, next); continue;
        case DEC: next = reduce_dup_dec(prev, next); continue;
        default: break;
      }
      case MAT:
      case IFL:
      case SWI: switch (tag) {
        case ERA: next = reduce_mat_era(prev, next); continue;
        case LAM: next = reduce_mat_lam(prev, next); continue;
        case SUP: next = reduce_mat_sup(prev, next); continue;
        case CTR: next = reduce_mat_ctr(prev, next); continue;
        case W32: case CHR: next = reduce_mat_w32(prev, next); continue;
        case INC: next = reduce_mat_inc(prev, next); continue;
        case DEC: next = reduce_mat_dec(prev, next); continue;
        default: break;
      }
      case OPX: switch (tag) {
        case ERA: next = reduce_opx_era(prev, next); continue;
        case LAM: next = reduce_opx_lam(prev, next); continue;
        case SUP: next = reduce_opx_sup(prev, next); continue;
        case CTR: next = reduce_opx_ctr(prev, next); continue;
        case W32: case CHR: next = reduce_opx_w32(prev, next); continue;
        case INC: next = reduce_opx_inc(prev, next); continue;
        case DEC: next = reduce_opx_dec(prev, next); continue;
        default: break;
      }
      case OPY: switch (tag) {
        case ERA: next = reduce_opy_era(prev, next); continue;
        case LAM: next = reduce_opy_lam(prev, next); continue;
        case SUP: next = reduce_opy_sup(prev, next); continue;
        case CTR: next = reduce_opy_ctr(prev, next); continue;
        case W32: case CHR: next = reduce_opy_w32(prev, next); continue;
        case INC: next = reduce_opy_inc(prev, next); continue;
        case DEC: next = reduce_opy_dec(prev, next); continue;
        default: break;
      }
      default: break;
    }

    // No interaction: push term back to stack, update parent chain
    spush(prev, sbuf, &spos);
    while (spos > stop) {
      Term host = spop(sbuf, &spos);
      set(term_loc(host) + 0, next);
      next = host;
    }
    *HVM.spos = spos;
    return next;
  }
}

Term reduce(Term term) {
  return reduce_impl(term, false);
}

Term reduce_owned(Term term) {
  return reduce_impl(term, true);
}

Term reduce_at(Loc host) {
  Term tm0 = got(host);
  if (term_tag(tm0) >= ERA) return tm0;
  Term tm1 = reduce(tm0);
  set(host, tm1);
  return tm1;
}

// Consumes the term stored in `host` and returns its WHNF without writing it
// back. This is used only by compiled reuse: the caller owns the returned term,
// while the frame slot stays empty and can be reclaimed without retaining a
// second alias. Substituted VAR/DP cells are consumed atomically for the same
// reason; unresolved DUP cells remain shared until their interaction fires.
Term reduce_take_at(Loc host) {
  Term tm0 = take(host);
  if (term_tag(tm0) >= ERA) return tm0;
  return reduce_impl(tm0, true);
}

Term normal(Term term) {
  Term wnf = reduce(term);
  Tag tag = term_tag(wnf);
  Lab lab = term_lab(wnf);
  Loc loc = term_loc(wnf);
  switch (tag) {
    case LAM: { Term bod = got(loc + 0); bod = normal(bod); set(term_loc(wnf) + 0, bod); return wnf; }
    case APP: { Term fun = got(loc + 0); Term arg = got(loc + 1); fun = normal(fun); arg = normal(arg); set(term_loc(wnf) + 0, fun); set(term_loc(wnf) + 1, arg); return wnf; }
    case SUP: { Term tm0 = got(loc + 0); Term tm1 = got(loc + 1); tm0 = normal(tm0); tm1 = normal(tm1); set(term_loc(wnf) + 0, tm0); set(term_loc(wnf) + 1, tm1); return wnf; }
    case DP0:
    case DP1: { Term val = got(loc + 0); val = normal(val); set(term_loc(wnf) + 0, val); return wnf; }
    case CTR: { u64 cid = lab; u64 ari = HVM.cari[cid]; for (u64 i = 0; i < ari; i++) { Term arg = got(loc + i); arg = normal(arg); set(term_loc(wnf) + i, arg); } return wnf; }
    case MAT:
    case IFL:
    case SWI: { u64 len = tag == SWI ? lab : tag == IFL ? 2 : HVM.clen[lab]; for (u64 i = 0; i <= len; i++) { Term arg = got(loc + i); arg = normal(arg); set(term_loc(wnf) + i, arg); } return wnf; }
    default: return wnf;
  }
}
